File: cracking-the-code-interview/VII_technical-questions/technique2/fibonacci.py
def fibonacci(n):
    if n < 2:
        return n

    return fibonacci(n - 1) + fibonacci(n - 2)

print("fibonacci(10)", fibonacci(10))

# Memoized
c = {}

def memoized(n):
    if n < 2:
        return n

    if c.get(n):
        return c.get(n)
    else:
        c[n] = memoized(n - 1) + memoized(n - 2)

    return c[n]

print("memoized(300)", memoized(300))

File: go/fibonacci.go
package main

import (
	"fmt"
)

// fibonacci is a function that returns
// a function that returns an int.
func fibonacci() func() int {
	a, b := 1, 0
	return func() int {
		a, b = b, a+b
		//		fmt.Fprintf(os.Stdout, "a: %d, b: %d\n", a, b)
		return a
	}
}

func main() {
	f := fibonacci()
	for i := 0; i < 10; i++ {
		fmt.Println(f())
	}
}
